### PR TITLE
[Lang] Build csr sparse matrix on GPU using coo format ndarray

### PR DESCRIPTION
**Pull request by FantasyVR**
Related issue = https://github.com/taichi-dev/taichi/issues/2906

It's more convenient for the users to generate COO format sparse matrix. But it's more efficient for the computer to use CSR format sparse matrix. In this pr, we provide a API to construct a CSR format sparse matrix using COO format input ndarrays.

### CODE REVIEW

File: python/taichi/linalg/sparse_matrix.py [hunk 1]
```diff
@@ -198,28 +198,29 @@ def build_from_ndarray(self, ndarray):
...
+    def build_coo(self, data, row_idx, col_idx):
+        """Build a CSR format sparse matrix from COO format inputs.
 
         Args:
-            data (ti.ndarray): CSR format data array of the matrix.
-            indices (ti.ndarray): CSR format index array of the matrix.
-            indptr (ti.ndarray): CSR format index pointer array of the matrix.
+            data (ti.ndarray): the entries of the matrix.
+            indices (ti.ndarray): the row indices of the matrix entries.
+            indptr (ti.ndarray): the column indices of the matrix entries.
```
Comment (strongoier) [inline python/taichi/linalg/sparse_matrix.py:207]: The arg names seem incorrect.

File: python/taichi/linalg/sparse_matrix.py [hunk 2]
```diff
@@ -198,28 +198,29 @@ def build_from_ndarray(self, ndarray):
...
-                indices, Ndarray) or not isinstance(indptr, Ndarray):
+                row_idx, Ndarray) or not isinstance(col_idx, Ndarray):
             raise TaichiRuntimeError(
                 'Sparse matrix only supports building from [ti.ndarray, ti.Vector.ndarray, ti.Matrix.ndarray].'
             )
-        elif data.dtype != f32 or indices.dtype != i32 or indptr.dtype != i32:
+        elif data.dtype != f32 or row_idx.dtype != i32 or col_idx.dtype != i32:
             raise TaichiRuntimeError(
                 'Sparse matrix only supports building from float32 data and int32 indices/indptr.'
             )
```
Comment (strongoier) [inline python/taichi/linalg/sparse_matrix.py:220]: Is this error message still valid after you change the arg names?

**Review by strongoier — APPROVED** [review]
LGTM!

File: python/taichi/linalg/sparse_matrix.py (hunk 2 above)
Comment (FantasyVR) [reply python/taichi/linalg/sparse_matrix.py]: Thanks, the error message is still valid. I changed some content to make it more clear.

**Pull request merged**